src/schedule/schedule.py

```diff
@@ -1,27 +1,69 @@
-from abc import ABC, abstractmethod
+# pylint: disable=useless-parent-delegation
+from abc import abstractmethod
 
+import marshmallow as marsh
+from peewee import ForeignKeyField
 
-class Schedule(ABC):
+from src.base_model import BaseModel
+from src.schedule.schedule_strategy import ScheduleStrategy
+
+
+class Schedule(BaseModel):
     """An abstract schedule for spawning SUMO vehicles."""
 
-    @classmethod
+    class Schema(BaseModel.Schema):
+        """Schema for Schedule."""
+
+        def _make(self, data: dict) -> "BaseModel":
+            """Constructs a Schedule from a dictionary.
+
+            :param data: The dictionary.
+            :return: A Schedule.
+            """
+            return super()._make(data)
+
+        strategy_id = marsh.fields.UUID(required=True)
+
+    _blocked: bool
+    strategy_id = ForeignKeyField(ScheduleStrategy, null=False)
+
+    def __init__(self, *args, **kwargs):
+        """Constructs a Schedule."""
+        super().__init__(*args, **kwargs)
+        self._blocked = False
+
     @abstractmethod
-    def from_json(cls, json_object: str) -> "Schedule":
-        """Constructs a schedule from a JSON object.
+    def _spawn(self, traci_wrapper: "ITraCiWrapper"):
+        """Spawns a vehicle.
 
-        :param json_object: The JSON object.
-        :type json_object: str
-        :return: A schedule.
-        :rtype: Schedule
+        :param traci_wrapper: The TraCI wrapper to give the vehicle to.
         """
         raise NotImplementedError()
 
-    @abstractmethod
-    def maybe_spawn(self, tick: int):
-        """Determines whether a vehicle should be spawned at the
-        current tick and spawns it if so.
+    def maybe_spawn(self, tick: int, traci_wrapper: "ITraCiWrapper"):
+        """Spawns a vehicle if the schedule strategy allows it.
 
-        :param tick: The current tick.
-        :type tick: int
+        :param tick: The current tick
+        :param traci_wrapper: The TraCI wrapper to give the spawned vehicle to.
```

Comment:
Aren't you able to specify a type of a param in the docstrings somehow? 🤔

Reply:
Yes you are as far as I know we agreed to not put type information in the docstrings because it is redundant to the python type hints.